cpp/src/replace/replace.cu

```diff
@@ -225,29 +334,115 @@ struct replace_kernel_forwarder {
+      }
+    } else {
+      if (replacement_values.has_nulls()) {
+        replace_first = replace_strings_first_pass<false, true>;
+        replace_second = replace_strings_second_pass<false, true>;
+      } else {
+        replace_first = replace_strings_first_pass<false, false>;
+        replace_second = replace_strings_second_pass<false, false>;
+      }
+    }
+
+    // Create new offsets column to use in kernel
+    std::unique_ptr<cudf::column> offsets = cudf::make_numeric_column(cudf::data_type(cudf::type_id::INT32),
+                                                                      input_col.size() + 1,
+                                                                      cudf::mask_state::UNALLOCATED,
+                                                                      stream,
+                                                                      mr);
+    std::unique_ptr<cudf::column> indices = cudf::make_numeric_column(cudf::data_type(cudf::type_id::INT32),
+                                                                      input_col.size(),
+                                                                      cudf::mask_state::UNALLOCATED,
+                                                                      stream);
+
+    auto offsets_view = offsets->mutable_view();
+    auto indices_view = indices->mutable_view();
+
+    auto device_in = cudf::column_device_view::create(input_col);
+    auto device_values_to_replace = cudf::column_device_view::create(values_to_replace);
+    auto device_replacement = cudf::column_device_view::create(replacement_values);
+    auto device_offsets = cudf::mutable_column_device_view::create(offsets_view);
+    auto device_indices = cudf::mutable_column_device_view::create(indices_view);
+
+    // Call first pass kernel to get sizes in offsets
+    cudf::experimental::detail::grid_1d grid { input_col.size(), BLOCK_SIZE, 1 };
+    replace_first<<<grid.num_blocks, BLOCK_SIZE, 0, stream>>>(*device_in,
+                                                              *device_values_to_replace,
+                                                              *device_replacement,
+                                                              *device_offsets,
+                                                              *device_indices);
+
+    // Compute the offsets from the sizes
+    int32_t * offsets_data = offsets_view.data<int32_t>();
+    thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           offsets_data,
+                           offsets_data + offsets_view.size(),
+                           offsets_data);
+    int32_t size;
+    CUDA_TRY(cudaMemcpyAsync(&size, offsets_data + offsets_view.size() - 1, sizeof(int32_t), cudaMemcpyDefault, stream));
+
+    // Allocate chars array and output null mask
+    std::unique_ptr<cudf::column> output_chars =
+        cudf::strings::detail::create_chars_child_column(input_col.size(), 0, size, mr, stream);
```

Comment:
The null-count parameter here must be accurate for the output strings column and not the chars column in order for this to work properly. This may mean you need to compute the nulls first.